sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -686,74 +678,65 @@ public void processElement(ProcessContext context) throws IOException {
+        FileResultCoder.of(shardedWindowCoder, destinationCoder);
 
+    PCollection<FileResult<DestinationT>> results;
     if (computeNumShards == null && numShardsProvider == null) {
-      numShardsView = null;
       TupleTag<FileResult<DestinationT>> writtenRecordsTag =
           new TupleTag<>("writtenRecordsTag");
-      TupleTag<KV<ShardedKey<Integer>, UserT>> unwrittedRecordsTag =
-          new TupleTag<>("unwrittenRecordsTag");
+      TupleTag<KV<ShardedKey<Integer>, UserT>> spilledRecordsTag =
+          new TupleTag<>("spilledRecordsTag");
       String writeName = windowedWrites ? "WriteWindowedBundles" : "WriteBundles";
       PCollectionTuple writeTuple =
           input.apply(
               writeName,
-              ParDo.of(new WriteBundles(unwrittedRecordsTag, destinationCoder))
+              ParDo.of(new WriteBundles(spilledRecordsTag, destinationCoder))
                   .withSideInputs(sideInputs)
-                  .withOutputTags(writtenRecordsTag, TupleTagList.of(unwrittedRecordsTag)));
+                  .withOutputTags(writtenRecordsTag, TupleTagList.of(spilledRecordsTag)));
       PCollection<FileResult<DestinationT>> writtenBundleFiles =
-          writeTuple
-              .get(writtenRecordsTag)
-              .setCoder(FileResultCoder.of(shardedWindowCoder, destinationCoder));
+          writeTuple.get(writtenRecordsTag).setCoder(fileResultCoder);
       // Any "spilled" elements are written using WriteShardedBundles. Assign shard numbers in
       // finalize to stay consistent with what WriteWindowedBundles does.
-      PCollection<FileResult<DestinationT>> writtenGroupedFiles =
+      PCollection<FileResult<DestinationT>> writtenSpilledFiles =
           writeTuple
-              .get(unwrittedRecordsTag)
+              .get(spilledRecordsTag)
               .setCoder(KvCoder.of(ShardedKeyCoder.of(VarIntCoder.of()), input.getCoder()))
-              .apply("GroupUnwritten", GroupByKey.<ShardedKey<Integer>, UserT>create())
+              // Here we group by a synthetic shard number in the range [0, spill factor),
+              // just for the sake of getting some parallelism within each destination when
+              // writing the spilled records, whereas the non-spilled records don't have a shard
+              // number assigned at all. Drop the shard number on the spilled records so that
+              // shard numbers are assigned together to both the spilled and non-spilled files in
+              // finalize.
+              .apply("GroupSpilled", GroupByKey.<ShardedKey<Integer>, UserT>create())
               .apply(
-                  "WriteUnwritten",
-                  ParDo.of(new WriteShardedBundles(ShardAssignment.ASSIGN_IN_FINALIZE))
-                      .withSideInputs(sideInputs))
-              .setCoder(FileResultCoder.of(shardedWindowCoder, destinationCoder));
+                  "WriteSpilled", ParDo.of(new WriteShardedBundles()).withSideInputs(sideInputs))
```

Comment:
so you're now going to assign shard #s based on the key? This is potentially broken for batch, as some of the files (the ones written bye WriteWindowedBundles) are assigned in finalize, but the spilled files are assigned here. I worry that could lead to collisions.